src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -94,18 +52,544 @@ impl fmt::Display for Mode {
+        let fn_ty = callee.ty(cx.mir, cx.tcx);
+        match fn_ty.sty {
+            ty::FnDef(def_id, _) => {
+                match cx.tcx.fn_sig(def_id).abi() {
+                    Abi::RustIntrinsic |
+                    Abi::PlatformIntrinsic => {
+                        assert!(!cx.tcx.is_const_fn(def_id));
+                        match &cx.tcx.item_name(def_id).as_str()[..] {
+                            | "size_of"
+                            | "min_align_of"
+                            | "needs_drop"
+                            | "type_id"
+                            | "bswap"
+                            | "bitreverse"
+                            | "ctpop"
+                            | "cttz"
+                            | "cttz_nonzero"
+                            | "ctlz"
+                            | "ctlz_nonzero"
+                            | "overflowing_add"
+                            | "overflowing_sub"
+                            | "overflowing_mul"
+                            | "unchecked_shl"
+                            | "unchecked_shr"
+                            | "rotate_left"
+                            | "rotate_right"
+                            | "add_with_overflow"
+                            | "sub_with_overflow"
+                            | "mul_with_overflow"
+                            | "saturating_add"
+                            | "saturating_sub"
+                            | "transmute"
+                            => return true,
+
+                            _ => {}
+                        }
+                    }
+                    _ => {
+                        let is_const_fn =
+                            cx.tcx.is_const_fn(def_id) ||
+                            cx.tcx.is_unstable_const_fn(def_id).is_some() ||
+                            cx.is_const_panic_fn(def_id);
+                        if !is_const_fn {
+                            return true;
+                        }
+                    }
+                }
+            }
+            _ => return true,
+        }
+
+        Self::in_operand(cx, callee) || args.iter().any(|arg| Self::in_operand(cx, arg))
+    }
+}
+
+// Refers to temporaries which cannot be promoted as
+// promote_consts decided they weren't simple enough.
+struct IsNotPromotable;
+
+impl Qualif for IsNotPromotable {
+    const IDX: usize = 3;
+
+    fn in_call(
+        cx: &ConstCx<'_, 'tcx>,
+        callee: &Operand<'tcx>,
+        _args: &[Operand<'tcx>],
+        _return_ty: Ty<'tcx>,
+    ) -> bool {
+        if cx.mode == Mode::Fn {
+            if let ty::FnDef(def_id, _) = callee.ty(cx.mir, cx.tcx).sty {
+                // Never promote runtime `const fn` calls of
+                // functions without `#[rustc_promotable]`.
+                if !cx.tcx.is_promotable_const_fn(def_id) {
+                    return true;
+                }
+            }
+        }
+
+        // FIXME(eddyb) do we need "not promotable" in anything
```

Comment:
Nope, we already stabilized promoting anything in a const fn:

```rust
const fn foo() -> &'static u64 { &bar() }
const fn bar() -> u64 { 0x420_b1a23_17 }
```

Even user defined const fn calls, which is not promotable in `Mode::Fn`

And I can't think of anything else where "not promotable" is required outside `Mode::Fn`

Reply:
So... I think we can just merge the `IsNotPromotable` logic into `IsNotConst`? We still need it just to track the difference between `not const` and `not promotable` in `const fn`

Reply:
https://github.com/rust-lang/rust/pull/58403#discussion_r256277080